pkg/registry/batch/job/strategy_test.go

```diff
@@ -760,56 +729,659 @@ func TestJobStatusStrategy(t *testing.T) {
+					ResourceVersion: "0",
+					Generation:      0,
+				},
+				Spec: batch.JobSpec{
+					Selector:       validSelector,
+					Template:       validPodTemplateSpec,
+					ManualSelector: pointer.BoolPtr(true),
+					Parallelism:    pointer.Int32Ptr(1),
+				},
+			},
+		},
+		"force validation failure in pod template": {
+			job: &batch.Job{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            "myjob",
+					Namespace:       metav1.NamespaceDefault,
+					ResourceVersion: "0",
+					Generation:      1,
+				},
+				Spec: batch.JobSpec{
+					Selector: validSelector,
+					Template: api.PodTemplateSpec{
+						Spec: api.PodSpec{Volumes: []api.Volume{{Name: "volume-name"}, {Name: "volume-name"}}},
+					}, ManualSelector: pointer.BoolPtr(true),
+					Parallelism: pointer.Int32Ptr(1),
+				},
+			},
+
+			oldJob: &batch.Job{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            "myjob",
+					Namespace:       metav1.NamespaceDefault,
+					ResourceVersion: "0",
+					Generation:      0,
+				},
+				Spec: batch.JobSpec{
+					Selector:       validSelector,
+					Template:       validPodTemplateSpec,
+					ManualSelector: pointer.BoolPtr(true),
+					Parallelism:    pointer.Int32Ptr(1),
+				},
+			},
+			wantWarningsCount: 1,
+		},
 	}
-	if *newJob.Spec.Parallelism != 10 {
-		t.Errorf("Job status updates must now allow changes to job spec")
+	for val, tc := range cases {
+		t.Run(val, func(t *testing.T) {
+			gotWarnings := Strategy.WarningsOnUpdate(ctx, tc.job, tc.oldJob)
+			if len(gotWarnings) != int(tc.wantWarningsCount) {
+				t.Errorf("got warning length of %d but expected %d", len(gotWarnings), tc.wantWarningsCount)
+			}
+		})
 	}
-	errs := StatusStrategy.ValidateUpdate(ctx, newJob, oldJob)
-	if len(errs) != 0 {
-		t.Errorf("Unexpected error %v", errs)
+}
+func TestJobStrategy_WarningsOnCreate(t *testing.T) {
+	ctx := genericapirequest.NewDefaultContext()
+
+	theUID := types.UID("1a2b3c4d5e6f7g8h9i0k")
+	validSelector := &metav1.LabelSelector{
+		MatchLabels: map[string]string{"a": "b"},
+	}
+
+	job := &batch.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "myjob2",
+			Namespace: metav1.NamespaceDefault,
+			UID:       theUID,
+		},
+		Spec: batch.JobSpec{
+			Selector: nil,
+			Template: api.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: validSelector.MatchLabels,
+				},
+				Spec: api.PodSpec{
+					RestartPolicy: api.RestartPolicyOnFailure,
+					DNSPolicy:     api.DNSClusterFirst,
+					Containers:    []api.Container{{Name: "abc", Image: "image", ImagePullPolicy: "IfNotPresent", TerminationMessagePolicy: api.TerminationMessageReadFile}},
+				},
+			}},
+	}
+
+	testcases := map[string]struct {
+		jobName           string
+		wantWarningsCount int32
+	}{
+		"happy path job": {
+			jobName: "myjob2",
+		},
+		"dns invalid name": {
+			wantWarningsCount: 1,
+			jobName:           "my job2",
+		},
+	}
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			job.Name = tc.jobName
```

Comment:
I'm not sure about mutating the shared `job` object here. What is both test cases run in parallel? I would suggest instead to declare validPodSpec as in other tests.

Reply:
Good idea! I’ll do that on Monday! Thank you for your time!

Reply:
I added this change.